akka-projection-poc2/src/main/scala/akka/projection/eventsourced/jdbc/JdbcProjectionHandler.scala

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+package akka.projection.eventsourced.jdbc
+
+import java.sql.Connection
+
+import scala.collection.immutable
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+
+import akka.Done
+import akka.persistence.query.Offset
+import akka.persistence.query.Sequence
+import akka.projection.scaladsl.AbstractGroupedEventsHandler
+import akka.projection.scaladsl.AbstractSingleEventHandler
+import akka.projection.scaladsl.OffsetManagedByProjectionHandler
+
+abstract class JdbcSingleEventHandlerWithTxOffset[Event](eventProcessorId: String, tag: String)(
+    implicit ec: ExecutionContext)
+    extends JdbcProjectionHandlerWithTxOffset[Offset](eventProcessorId, tag)(ec)
+    with AbstractSingleEventHandler[Event] {
+
+  override def onEvent(event: Event): Future[Done]
+
+}
+
+abstract class JdbcGroupedEventsHandlerWithTxOffset[Event](
+    eventProcessorId: String,
+    tag: String,
+    override val n: Int,
+    override val d: FiniteDuration)(implicit ec: ExecutionContext)
+    extends JdbcProjectionHandlerWithTxOffset[Offset](eventProcessorId, tag)(ec)
+    with AbstractGroupedEventsHandler[Event] {
+
+  def onEvents(events: immutable.Seq[Event]): Future[Done]
+
+}
+
+abstract class JdbcProjectionHandlerWithTxOffset[Event](eventProcessorId: String, tag: String)(
+    implicit val ec: ExecutionContext)
+    extends OffsetManagedByProjectionHandler[Offset] {
+
+  private var currentOffset: Offset = Offset.noOffset
+
+  def getConnection(): Connection
+
+  override def readOffset(): Future[Option[Offset]] = {
+    val c = getConnection()
+    val pstmt =
+      c.prepareStatement("SELECT offset FROM akka_cqrs_sample.offsetStore WHERE eventProcessorId = ? AND tag = ?")
+    pstmt.setString(1, eventProcessorId)
+    pstmt.setString(2, tag)
+    val rs = pstmt.executeQuery()
+    try {
+      if (rs.next()) {
+        // FIXME other types of offsets
+        Future.successful(Some(Offset.sequence(rs.getLong(1))))
+      } else {
+        Future.successful(None)
+      }
+    } finally {
+      // FIXME better resource closing
+      rs.close()
+      pstmt.close()
+      c.close()
+    }
+  }
+
+  override final def setCurrentOffset(offset: Offset): Unit = {
+    currentOffset = offset
+  }
```

Comment:
I kind of hide the Offset from the user with this. So it's not required in the signature of `onEvent` and `saveOffset` can be called without passing on the Offset.

It's maybe better to make the Offset visible in `onEvent` of OffsetManagedByProjectionHandler. I can try that.

Also, `EventEnvelope` isn't a great type because `event: Any`. I have an idea of how we can introduce a type there in Akka and still be source and binary compatible. Alternatively we have to create a new `EventEnvelope[Event]` in akka-projection.

Reply:
changed in 724ce60